connectors/sources/outlook.py

```diff
@@ -0,0 +1,944 @@
+
+            user_account = Account(
+                primary_smtp_address=user.get("attributes", {}).get("mail"),
+                config=configuration,
+                access_type=IMPERSONATION,
+            )
+            yield user_account
+
+
+class Office365Users:
+    """Fetch users from Office365 Active Directory"""
+
+    def __init__(self, client_id, client_secret, tenant_id):
+        self.tenant_id = tenant_id
+        self.client_id = client_id
+        self.client_secret = client_secret
+
+    @cached_property
+    def _get_session(self):
+        return aiohttp.ClientSession(raise_for_status=True)
+
+    def _check_errors(self, response):
+        match response.status:
+            case 400:
+                raise UnauthorizedException("Found invalid tenant id value")
+            case 401:
+                raise UnauthorizedException(
+                    "Found invalid client id or client secret value"
+                )
+            case 403:
+                raise Forbidden(
+                    f"Missing permission or something went wrong. Error: {response}"
+                )
+            case 404:
+                raise NotFound(f"Resource Not Found. Error: {response}")
+            case _:
+                raise
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=UnauthorizedException,
+    )
+    async def _fetch_token(self):
+        try:
+            async with self._get_session.post(
+                url=f"https://login.microsoftonline.com/{self.tenant_id}/oauth2/v2.0/token",
+                data={
+                    "grant_type": "client_credentials",
+                    "client_id": self.client_id,
+                    "client_secret": self.client_secret,
+                    "scope": API_SCOPE,
+                },
+            ) as response:
+                token_response = await response.json()
+                return token_response["access_token"]
+        except Exception as exception:
+            self._check_errors(response=exception)
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def get_users(self):
+        try:
+            access_token = await self._fetch_token()
+            async with self._get_session.get(
+                url="https://graph.microsoft.com/v1.0/users",
+                headers={
+                    "Authorization": f"Bearer {access_token}",
+                    "Content-Type": "application/json",
+                },
+            ) as response:
+                json_response = await response.json()
+                users_mails = [
+                    user.get("mail")
+                    for user in json_response.get("value")
+                    if user.get("mail")
+                ]
+                for user in users_mails:
+                    yield user
```

Comment:
This can be simplified to:
```suggestion
                for user in json_response.get("value", []):
                    if "mail" in user:
                        yield user.get("mail")
```

But I think the methods in `Client` class should be general. If you name the method name `get_users`, you should simply yield `user`, and let the caller handle it as appropriate.

Reply:
Agreed 👍